pyretailscience/analysis/product_association.py

```diff
@@ -226,83 +203,121 @@ def _calc_association(  # noqa: C901 (ignore complexity) - Excluded due to min_*
         if min_uplift < 0.0:
             raise ValueError("Minimum uplift must be greater or equal to 0.")
 
-        unique_combo_df = df[[group_col, value_col]].drop_duplicates()
-        unique_combo_df[value_col] = pd.Categorical(unique_combo_df[value_col], ordered=True)
-        unique_combo_df[group_col] = pd.Categorical(unique_combo_df[group_col], ordered=True)
-
-        sparse_matrix = csc_matrix(
-            (
-                [1] * len(unique_combo_df),
-                (
-                    unique_combo_df[group_col].cat.codes,
-                    unique_combo_df[value_col].cat.codes,
-                ),
-            ),
+        if isinstance(df, pd.DataFrame):
+            df = ibis.memtable(df)
+
+        unique_transactions = df
+        total_transactions = unique_transactions.alias("t")[group_col].nunique().name("total_count")
+
+        product_occurrences = (
+            unique_transactions.group_by(value_col)
+            .aggregate(
+                occurrences=lambda t: t[group_col].nunique(),
+            )
+            .mutate(occurrence_probability=lambda t: t.occurrences / total_transactions)
+            .filter(lambda t: t.occurrences >= min_occurrences)
         )
 
-        row_count = sparse_matrix.shape[0]
+        left_table = unique_transactions.rename({"item_1": value_col})
+        right_table = unique_transactions.rename({"item_2": value_col})
+
+        merged_df = left_table.join(
+            right_table,
+            predicates=[
+                left_table[group_col] == right_table[group_col],
+                left_table["item_1"] < right_table["item_2"],
+            ],
+            lname="",
+            rname="{name}_right",
+        )
```

Comment:
If target_item is none, you can implement that here to improve performance. Something like the below? That would drastically cut down the size of the join.
```suggestion
        join_logic = [left_table[group_col] == right_table[group_col]]
        if target_item is None:
            join_logic.append(left_table["item_1"] < right_table["item_2"])
        else:
           join_logic.extend([
               left_table["item_1"] != right_table["item_2"],
               left_table["item_1"] == target_item,
          ])
        merged_df = left_table.join(
            right_table,
            predicates=join_logic,
            lname="",
            rname="{name}_right",
        )
```